### ISSUE TITLE
freeCodeCamp JavaScript Console not responding appropriately

### ISSUE THREAD

**Issue by tknick2** [freeCodeCamp/freeCodeCamp#38474]
**Describe your problem and how to reproduce it:**
"Challenge: Understanding the Differences between the freeCodeCamp and Browser Console" indicates that the freeCodeCamp console should display the string twice, but it only displays once, even when the challenge has been completed.

**Add a Link to the page with the problem:**

**Tell us about your browser and operating system:**
* Browser Name: 
* Browser Version: 
* Operating System: 

**If possible, add a screenshot here (you can drag and drop, png, jpg, gif, etc. in this box):**

**Comment by moT01** [comment]
The was written before our console was changed to behave more accurately. This may need a slight rewrite, or possibly just remove this challenge.

**Comment by Sky020** [comment]
> or possibly just remove this challenge.

I have seen far too many campers stuck on challenges, because they rely on the fCC console entirely for debugging. So, I would suggest we just rewrite part of it. Perhaps, putting information/emphasis on how to effectively use the dev-tools.

**Comment by ojeytonwilliams** [comment]
@Sky020 do you know what it's not doing properly?  I can see about improving it, if it's unhelpful.

**Comment by ruslan-akhm** [comment]
It actually works the way it supposed to for me.
I believe it's about unclear explanation. 
The thing is, if you type: 
`let output = "Get this to log once in the browser console and twice in the freeCodeCamp console";

console.log(output);
console.clear();

console.log(output);`

it will show two outputs in FCC console and only one in browser (I'm using Chrome)

**Comment by Sky020** [comment]
> @Sky020 do you know what it's not doing properly? I can see about improving it, if it's unhelpful.

@ojeytonwilliams , users struggle with the recursive challenges, because they will `console.log` their final output (or similar), and it will match the tests, but the tests will not pass, because the actual output is multiple arrays. Something like this:
```
// Test: myFunc(arr) should return [0, 1, 2, 3]
// console.log to fCC
[0, 1, 2, 3]
// console.log to console (dev-tools)
[0]
[0, 1]
[0, 1, 2]
[0, 1, 2, 3]
```
I will try to find an actual example, at some point.

**Comment by ojeytonwilliams** [comment]
An example would be great, thanks.  For what it's worth, I tried a recursion challenge and it's behaving as expected:

```js
function countdown(n){
  if(n < 1) return [];
  const countArray = countdown(n-1);
  countArray.unshift(n);
  console.log(n, countArray)
  return countArray;
}
countdown(3);
```
gives 

```
1 [ 1 ]
2 [ 2, 1 ]
3 [ 3, 2, 1 ]
```
in the fCC console and

```
1 [1]
2 (2) [2, 1]
3 (3) [3, 2, 1]
```

in the dev console.  Not identical, but I think it's okay.

**Comment by moT01** [comment]
After a little testing, it looks like `log` is the only console method that works in our console as far as I can tell. All the other ones don't seem to be working. So the `clear` method in this challenge won't work as @ruslan-akhm mentioned. Perhaps we could look into adding the ability to use some of the other methods.

**Comment by Sky020** [comment]
> An example would be great, thanks.

@ojeytonwilliams , I have not found any examples. Potentially, it has been resolved, but only the incorrect code produced it. So, users tests were failing, due to multiple arrays being returned, instead of one, but the users thought their code was correct because the console only should the final step. Something like that.

**Comment by ojeytonwilliams** [comment]
@Sky020 thanks for looking into this.  If you do notice situations where it's not working as it should, we can always revisit this.

@moT01 log is the only one that's proxied to the fcc console.  We could do the others, but I'm not convinced it's that valuable.  Perhaps just a note in the challenge like 

> other log levels (info, warn, error and debug) do not appear in the freeCodeCamp console.

would be enough.

**Comment by moT01** [comment]
I think this challenge should be rewritten to explain that `log` is the only method available in our console. Right now, it's explaining a problem that doesn't exist anymore. So if we don't change it to be about that, I'm not sure what else we could make it about - or what it's purpose would be.

**Comment by lasjorg** [comment]
Not sure if this is the correct issue for this question/observation. It just seemed like this issue now is about the logging in general, so I will mention it anyway. The ES6 challenges that deal with import/export are throwing errors in the console.

For example:
https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/es6/reuse-javascript-code-using-import

Throws `ReferenceError: require is not defined` in the console.


I don't really know much about the log code but it looks like the eval on line 73 in test-evaluator.js is throwing. I see

`var _string_functions = require("string_functions.js");`

inside the `e.data.build` property.

If this isn't the right place for this, or we need to open a separate issue for it, just let me know.

**Comment by ojeytonwilliams** [comment]
@lasjorg yeah, this is a good place to discuss it, thanks for bringing it up.  What's happening is Babel transpiles `import` statements to use `require` instead:

```js
import { uppercaseString } from './math_functions.js'
// becomes
var _math_functions = require("./math_functions.js");
```
but `require` does not exist in the browser, hence the `ReferenceError` you saw.

We cannot just define `require` in the `js-setup`, because that gets transpiled to `_require`. However, what we can do is define a fake `require` inside the `test-evaluator` that returns whatever should have been returned.  @RandellDawson what do you think?

**Comment by Sky020** [comment]
> Any recent forum posts you can find with this problem @Sky020 ?

@moT01 @ojeytonwilliams 
I think this is was the issue: https://www.freecodecamp.org/forum/t/basic-javascript-use-recursion-to-create-a-countdown/322222
The challenge has, since, been changed. I do not think it applies any more. Maybe `ieahleen` or @ArielLeslie remember what I have been referring to.

**Comment by Sky020** [comment]
@ojeytonwilliams , This just recently popped up on the forum:
https://www.freecodecamp.org/forum/t/doubt-in-basic-of-java-script/369825
The console seems to show that the function is self-invoking.
```
function myLocalScope() {
 'use strict';
  var myVar=5;
  console.log(myVar);
}
myLocalScope();
```
Here is the related challenge: https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/local-scope-and-functions

**Comment by lasjorg** [comment]
I think that is comming from the setup/teardown code.
https://github.com/freeCodeCamp/freeCodeCamp/blob/master/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/local-scope-and-functions.english.md

**Comment by Ieahleen** [comment]
> > An example would be great, thanks.
> 
> @ojeytonwilliams , I have not found any examples. Potentially, it has been resolved, but only the incorrect code produced it. So, users tests were failing, due to multiple arrays being returned, instead of one, but the users thought their code was correct because the console only should the final step. Something like that.

there were a lot of issues when the console would print just a comma separated list of elements for arrays, instead of arrays, so it was impossible to distinguish between flat arrays and multidimensional ones.

RIght now I have seen:
 - not understanding that the functions are called by the tests, so confusion in seeing multiple lines in the console for each console.log inside the function
 - mismatch between challenges describing how the console work and how it works.

Most of the issues seems resolved with last update of the console.

**Comment by ojeytonwilliams** [comment]
> This just recently popped up on the forum:
https://www.freecodecamp.org/forum/t/doubt-in-basic-of-java-script/369825
The console seems to show that the function is self-invoking.

Thanks, that's definitely confusing behaviour.  

> I think that is comming from the setup/teardown code.

I think so too, the after test is the problem.  That challenge has a lot of redundant code - none of the setup or teardown is needed.  It would be fine to simply invoke `myLocalScope()` in a test.

> there were a lot of issues when the console would print just a comma separated list of elements for arrays, instead of arrays, so it was impossible to distinguish between flat arrays and multidimensional ones.

Just to confirm, @Ieahleen, that's no longer happening?

Anyway, I feel like specific challenges need work (the challenge @Sky020 found and the one that introduces the fCC console), but that the console as a whole is basically okay.  Is that what everyone else thinks?

**Comment by Ieahleen** [comment]
yes, at this time the console print arrays as expected. I have not checked
recently for objects (that also can cause confusion if objects are printed
in an unclear way, but basing on the array thing, that may also be ok)

**Comment by Sky020** [comment]
>Anyway, I feel like specific challenges need work (the challenge @Sky020 found and the one that introduces the fCC console), but that the console as a whole is basically okay. Is that what everyone else thinks?

To add to the issue with [this challenge](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/local-scope-and-functions): You cannot pass the challenge with this code, because it uses `let` instead of `var`.
```
function myLocalScope() {
  'use strict';
  // Only change code below this line
  let myVar = 1;
  console.log(myVar);
}
myLocalScope();
// Run and check the console
// myVar is not defined outside of myLocalScope
// Now remove the console log line to pass the test
```
I realise this is only an issue for those who are familiar with JavaScript outside of the prior fCC lessons, because otherwise users will not know about `let` and `const`. But, I encountered fCC after learning to declare using `let`.

**Comment by lasjorg** [comment]
@ojeytonwilliams We don't actually need to run the function in the test, right? The check for the local variable is done using a regex and `myVar` will be undefined as long as it was not declared _and_ assigned outside the function.

The current test does not check that the variable is not declared, even though it is worded that way in the challenge. It checks that it was not assigned a value. You can pass the test with a `myVar` variable outside the function as long as it is not assigned a value.

I'm not sure if the wording in the challenge needs to be changed or the test?

**Comment by ojeytonwilliams** [comment]
@Sky020 good point, it should accept all three.
@lasjorg you are right. I was thinking we needed to call `localScope` and then check if `myVar` had been assigned inside `localScope`, but clearly not.

> The current test does not check that the variable is not declared, even though it is worded that way in the challenge.

Yep, the test is not quite right, I'll take a look.

Would anyone be interested in improving the [console challenge](https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/debugging/understanding-the-differences-between-the-freecodecamp-and-browser-console)?

**Comment by Sky020** [comment]
>Would anyone be interested in improving the console challenge?

@ojeytonwilliams , I would be interested in improving this. Should I open a new issue to get a discussion going?

**Comment by ojeytonwilliams** [comment]
@Sky020 discussing it here seems fine, the right people should get notified.  Also, if you have some ideas about how to improve it, feel free to create the PR and we can discuss further once that's up.

**Comment by lasjorg** [comment]
These two challenge has the camper using a console.log, but nothing seems to get logged to the fCC console, only to the browser console.

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/es6/handle-a-fulfilled-promise-with-then
https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/es6/handle-a-rejected-promise-with-catch

**Comment by Sky020** [comment]
Reserrecting this with this conversation: https://forum.freecodecamp.org/t/a-strange-behavior-in-the-console-i-do-not-know-whether-it-is-a-bug-or-it-is-a-problem-of-my-browser/407951

Regarding this lesson: https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/es6/create-an-export-fallback-with-export-default

My thoughts:
>Perhaps this is something that can be looked into by catching that error, and throwing something more useful.

**Comment by ojeytonwilliams** [comment]
I think we should suppress import/export errors entirely, since they're never informative.
